bazel/repositories.bzl

```diff
@@ -1,532 +1,150 @@
+        #
+        # quiet = False,
     )
-
-def boringssl_repositories():
-    native.git_repository(
-        name = "boringssl",
-        commit = "bfd36df3da38dbf8828e712f42fbab2a0034bc40",  # 2017-02-02
-        remote = "https://boringssl.googlesource.com/boringssl",
+    if result.return_code != 0 or ctxt.attr.debug:
+        print("External dep build exited with return code: %d" % result.return_code)
+        print(result.stdout)
+        print(result.stderr)
+        if result.return_code != 0:
+            fail("External dep build failed")
+
+def envoy_dependencies(path = "@envoy_deps//", local_protobuf_bzl = None):
+    # Used only for protobuf.bzl.
+    if local_protobuf_bzl:
+        native.new_local_repository(
+            name = "protobuf_bzl",
+            path = "/thirdparty/protobuf-3.2.0",
+            # We only want protobuf.bzl, so don't support building out of this repo.
+            build_file_content = "",
+        )
+    else:
+        native.new_git_repository(
+            name = "protobuf_bzl",
+            # Using a non-canonical repository/branch here. This is a workaround to the lack of
+            # merge on https://github.com/google/protobuf/pull/2508, which is needed for supporting
+            # arbitrary CC compiler locations from the environment. The branch is
+            # https://github.com/htuch/protobuf/tree/v3.2.0-default-shell-env, which is the 3.2.0
+            # release with the above mentioned PR cherry picked.
+            commit = "d490587268931da78c942a6372ef57bb53db80da",
+            remote = "https://github.com/htuch/protobuf.git",
+            # We only want protobuf.bzl, so don't support building out of this repo.
+            build_file_content = "",
+        )
+
+    # Set this to True to make the build debug cycles faster.
+    debug_build = False
+
+    envoy_repository = repository_rule(
+        implementation = _repository_impl,
+        local = debug_build,
+        environ = ["CC", "CXX", "LD_LIBRARY_PATH"],
+        attrs = {"debug": attr.bool(default=False)},
     )
 
-def googletest_repositories():
-    BUILD = """
-cc_library(
-    name = "googletest",
-    srcs = [
-        "googlemock/src/gmock-all.cc",
-        "googletest/src/gtest-all.cc",
-    ],
-    hdrs = glob([
-        "googlemock/include/**/*.h",
-        "googlemock/src/*.cc",
-        "googletest/include/**/*.h",
-        "googletest/src/*.cc",
-        "googletest/src/*.h",
-    ]),
-    includes = [
-        "googlemock",
-        "googlemock/include",
-        "googletest",
-        "googletest/include",
-    ],
-    visibility = ["//visibility:public"],
-)
-"""
-    native.new_git_repository(
-        name = "googletest_git",
-        build_file_content = BUILD,
-        # v1.8.0 release
-        commit = "ec44c6c1675c25b9827aacd08c02433cccde7780",
-        remote = "https://github.com/google/googletest.git",
+    envoy_repository(
```

Comment:
I agree that we want to avoid consuming projects having to build all of Envoy's deps if they have their own versions. I've fixed this in the latest update, with a `skip_targets` option to `envoy_dependencies`. This list gets plumbed all the way down to the underlying `make` invocation, it's not just a bind omission anymore.

Having a monolithic dependency means that developers need to rebuild everything each time a dependency changes. However, this is rare. Much more frequent is the case that a developer checks out a new tree. So, I think we should optimize for the common case and make this process fast.

I've added code to the build system to profile how long each step in the build recipes takes, and to compare two different styles of build:

1. Build everything in parallel under recursive make, under a make jobserver with all CPUs available.

2. Build each dependency (via its build recipe) sequentially with all CPUs available at each build. This simulates what we would see if we had separate repository_rules for each dependency. 

The total time for (1) is 1m 49s. The total time for (2) is 3m 9s, almost 70% increase in build time and definitely noticeable.

To understand why it's slower, take a look at https://github.com/htuch/data-dump/tree/master/envoy-build-profiles. It's clear that the `configure` and `buildconf` steps take substantial time, and in (1) they overlap with each other and CPU intensive work, hiding the latency and in (2) they bottleneck and the machine is sitting mostly idle while they happen.

So, I think we should stick with the monolithic dep. 

As far as `repository_rule` goes and more general Bazel support, this is something that needs to be addressed foundationally. If `repository_rule`s execute sequentially, then the machine will be underutilized during the autoconf steps. If they execute in paralle, there is a need to basically act like a make jobserver to coordinate them.